### PR TITLE
Add support for specifying device tree for the Microsemi Expansion board

### PR DESCRIPTION
**Pull request by alistair23**
This adds support for the Microsemi expansion board. There seems to be a bug with how the DTB is copied that requires the third commit to work around.

### CODE REVIEW

**Comment by alistair23** [comment]
This addresses https://github.com/riscv/opensbi/issues/38 and probably https://github.com/riscv/opensbi/issues/35 as well

**Comment by olofj** [comment]
The slope is deep and slippery; do we really want opensbi to contain device trees for hardware support?

**Comment by alistair23** [comment]
This series really only adds the device tree, but I do see your point.

I think ideally we shouldn't, but currently for the HiFive Unleashed we don't really have a better option unfortunately. Once we have SMP boot with mainline u-boot (or at least mainline u-boot working) we can remove it, but until then I don't see much of an option.

I can add some comments saying that new boards shouldn't do this if that helps?

File: include/sbi/sbi_scratch.h [hunk 1]
```diff
@@ -57,6 +58,7 @@ struct sbi_scratch {
 	unsigned long hartid_to_scratch;
 	/** IPI type (or flags) */
 	unsigned long ipi_type;
+	unsigned long next_arg2;
 } __packed;
```
Comment (avpatel) [inline include/sbi/sbi_scratch.h]: Rename next_arg2 to prev_arg1. It is good to save Arg1 (a1 register) passed by previous booting stage in per-HART scratch space.

Also, define prev_arg1 before "next_arg1".

File: include/sbi/sbi_scratch.h [hunk 2]
```diff
@@ -30,6 +30,7 @@
 #define SBI_SCRATCH_HARTID_TO_SCRATCH_OFFSET	(7 * __SIZEOF_POINTER__)
 /** Offset of ipi_type member in sbi_scratch */
 #define SBI_SCRATCH_IPI_TYPE_OFFSET		(8 * __SIZEOF_POINTER__)
+#define SBI_SCRATCH_NEXT_ARG2_OFFSET		(9 * __SIZEOF_POINTER__)
 /** Maximum size of sbi_scratch */
```
Comment (avpatel) [inline include/sbi/sbi_scratch.h]: Rename SBI_SCRATCH_NEXT_ARG2_OFFSET to SBI_SCRATCH_PREV_ARG1_OFFSET

File: platform/sifive/fu540/platform.c [hunk 3]
```diff
@@ -53,11 +53,17 @@
 static void fu540_modify_dt(void *fdt)
 {
 	u32 i, size;
+	void *fdt_orig = ((void *)(sbi_scratch_thishart_ptr()->next_arg2));
 	int chosen_offset, err;
```
Comment (avpatel) [inline platform/sifive/fu540/platform.c]: Update as-per above comments.

File: firmware/fw_base.S [hunk 4]
```diff
@@ -169,6 +169,8 @@ _start_warm:
 	/* Note: fw_next_arg1() uses a0, a1, and ra */
 	call	fw_next_arg1
 	REG_S	a0, SBI_SCRATCH_NEXT_ARG1_OFFSET(tp)
+	call	fw_prev_arg1
+	REG_S	a0, SBI_SCRATCH_NEXT_ARG2_OFFSET(tp)
 	/* Note: fw_next_addr() uses a0, a1, and ra */
```
Comment (avpatel) [inline firmware/fw_base.S]: Save prev_arg1 before next_arg1.

File: include/sbi/sbi_scratch.h (hunk 1 above)
Comment (avpatel) [reply include/sbi/sbi_scratch.h]: If we don't have built-in DTB linked to OpenSBI then prev_arg1 will point to "a1" register passed by previous booting stage.

File: platform/sifive/fu540/platform.c [hunk 5]
```diff
@@ -69,10 +75,6 @@ static void fu540_modify_dt(void *fdt)
 			fdt_setprop_string(fdt, cpu_offset, "status", "masked");
 		memset(cpu_node, 0, sizeof(cpu_node));
 	}
-	size = fdt_totalsize(fdt);
-	err = fdt_open_into(fdt, fdt, size + 256);
-	if (err < 0)
-		sbi_printf("Device Tree can't be expanded to accmodate new node");
```
Comment (avpatel) [inline platform/sifive/fu540/platform.c]: Please add separate patch for changes in platform/sifive/fu540/platform.c

**Comment by alistair23** [comment]
@avpatel Do you think this is the right direction? I thought that the fw_base.S should handle copying the FDT and that there is a bug in there which is why this is needed. If we go down this road it won't be generic and each platform will need to handle copying the FDT. My ideal solution would be to fix the problem in fw_base.S that doesn't copy the FDT, just need to figure out what that is.

**Comment by atishp04** [comment]
I agree with Alistair. fdt_open_into will only work for platfroms that include libfdt. So we have to keep both fdt copying code in case we choose this route. We should fix it fw_base.S if possible.

**Comment by avpatel** [comment]
Like Olof already mentioned, this is a slippery slop. I gave more thought for alternate approach. Here's what I think:

Instead of the updating "struct sbi_scratch" and fw_base.S, we can do the following in fu540_modify_dt():
1. "fdt" parameter already points to original DTB (passed by previous stage or built-in DTB) copied over to final location
2. Get size of DTB size using libfdt APIs
3. Copy over DTB from "fdt" to "orig_fdt" where orig_fdt = fdt + <dtb_size> + <increase_in_dtb_size>
4. Now use fdt_open_into() with "fdt" as new location of DTB and "orig_fdt" as old location of DTB

**Comment by alistair23** [comment]
There seems to be a lot of confusion with what is going on.

I have pushed a v2 which fixes the original bug and removed the hack patch.

**Comment by alistair23** [comment]
V3, just updated the docs to make it explicit that in the future this will be removed when no longer needed.

**Pull request merged**